packages/engine/src/assets/state/ResourceState.ts

```diff
@@ -0,0 +1,246 @@
+    resources: {} as Record<string, Resource>
+  })
+})
+
+const getCurrentSizeOfResources = () => {
+  let size = 0
+  const resources = getState(ResourceState).resources
+  for (const key in resources) {
+    const resource = resources[key]
+    if (resource.metadata.size) size += resource.metadata.size
+  }
+
+  return size
+}
+
+const Callbacks = {
+  [ResourceType.GLTF]: {
+    onLoad: (response: GLTF, resource: State<Resource>) => {},
+    onProgress: (request: ProgressEvent, resource: State<Resource>) => {
+      resource.metadata.size.set(request.total)
+    },
+    onError: (event: ErrorEvent | Error, resource: State<Resource>) => {}
+  },
+  [ResourceType.Texture]: {
+    onLoad: (response: Texture, resource: State<Resource>) => {
+      if (response.mipmaps[0]) {
+        resource.metadata.size.set(response.mipmaps[0].data.length)
+      } else {
+        const height = response.image.height
+        const width = response.image.width
+        const size = width * height * 4
+        resource.metadata.size.set(size)
+      }
+    },
+    onProgress: (request: ProgressEvent, resource: State<Resource>) => {},
+    onError: (event: ErrorEvent | Error, resource: State<Resource>) => {}
+  }
+}
+
+const load = (
+  url: string,
+  resourceType: ResourceType,
+  entity: Entity,
+  args: LoadingArgs,
+  onLoad: (response: AssetType) => void,
+  onProgress: (request: ProgressEvent) => void,
+  onError: (event: ErrorEvent | Error) => void
+) => {
+  const resourceState = getMutableState(ResourceState)
+  const resources = resourceState.nested('resources')
+  if (!resources[url].value) {
+    resources.merge({
+      [url]: {
+        status: ResourceStatus.Unloaded,
+        type: resourceType,
+        references: [entity],
+        metadata: {},
+        onGPU: false
+      }
+    })
+  } else {
+    resources[url].references.merge([entity])
+  }
+
+  const resource = resources[url]
+  const callback = Callbacks[resourceType]
+  console.log('Resource Manager Loading Asset at: ' + url)
+  AssetLoader.load(
+    url,
+    args,
+    (response) => {
+      resource.status.set(ResourceStatus.Loaded)
+      resource.assetRef.set(response)
+      callback?.onLoad(response, resource)
+      onLoad(response)
+    },
+    (request) => {
+      resource.status.set(ResourceStatus.Loading)
+      callback?.onProgress(request, resource)
+      onProgress(request)
+    },
+    (error) => {
+      resource.status.set(ResourceStatus.Error)
+      callback?.onError(error, resource)
+      onError(error)
+    }
+  )
+}
+
+const unload = (url: string, resourceType: ResourceType, entity: Entity) => {
+  const resourceState = getMutableState(ResourceState)
+  const resources = resourceState.nested('resources')
+  if (!resources[url].value) {
+    console.error('ResourceManager:unload No resource exists for url: ' + url)
+    return
+  }
+
+  const resource = resources[url]
+
+  resource.references.set((entities) => {
+    const index = entities.indexOf(entity)
```

Comment:
We should be performing reference counting here; a single entity may use the same resource more than once.